test/test_unary_ufuncs.py

```diff
@@ -295,7 +317,41 @@ def test_reference_numerics(self, device, dtype, op):
                                            exact_dtype=exact_dtype, rtol=1e-3, atol=1e-2)
                     continue
 
-            self.assertEqualHelper(actual, expected, msg, dtype=dtype, exact_dtype=exact_dtype)
+            self.assertEqualHelper(actual, expected, msg, dtype=dtype, equal_nan=equal_nan, exact_dtype=exact_dtype)
+
+    # Tests that the function and its (array-accepting) reference produce the same
+    #   values on a range of tensors, including empty tensors, scalar tensors,
+    #   1D tensors and a large 2D tensor with interesting and extremal values
+    #   and discontiguities.
+    @suppress_warnings
+    @ops(unary_ufuncs)
+    def test_reference_numerics(self, device, dtype, op):
+        tensors = generate_numeric_tensors(device, dtype,
+                                           domain=op.domain)
+        self._test_reference_numerics(dtype, op, tensors)
+
+    @suppress_warnings
+    @ops(unary_ufuncs, allowed_dtypes=(*get_all_fp_dtypes(), *get_all_complex_dtypes(),
+                                       torch.int8, torch.int16, torch.int32, torch.int64))
+    def test_reference_numerics_hard(self, device, dtype, op):
+        if not op.handles_large_floats:
+            raise self.skipTest("This op does not handle large values")
+
+        tensors = generate_numeric_tensors_hard(device, dtype,
+                                                domain=op.domain)
+        self._test_reference_numerics(dtype, op, tensors)
+
+    @suppress_warnings
+    @ops(unary_ufuncs, allowed_dtypes=(*get_all_fp_dtypes(), *get_all_complex_dtypes()))
+    def test_reference_numerics_extremal(self, device, dtype, op):
+        handles_extremals = (op.handles_complex_extremals if
+                             dtype in (torch.cfloat, torch.cdouble) else op.handles_extremals)
+        if not handles_extremals:
+            raise self.skipTest("This op does not handle extremal values")
+
+        tensors = generate_numeric_tensors_extremal(device, dtype,
+                                                    domain=op.domain)
+        self._test_reference_numerics(dtype, op, tensors, "relaxed")
```

Comment:
Query for an OpInfo property for whether "relaxed" complex nan comparisons are needed.

Reply:
Don't we decide to enable `relaxed` for all unay complex compare on extreme? This is how I understand https://github.com/pytorch/pytorch/issues/50749. I thought you proposed that the test should be split into easy, hard, and extreme, and extreme will use "nan is nan". Because if we are adding a `relaxed_complex_nan` to `OpInfo`, why splitting the test is a dependency of fixing these two failures? We can read this flag in the original test without splitting it.

Reply:
The issue actually says we can consider it.

Splitting the tests is useful because it lets us focus on test granularity. If this nan comparison issue only happens on the extremal tests, for example, that's different from it happening on tiny complex numbers or large complex numbers. It would give us the option to skip just those extremal tests instead of skipping all of the historic test_reference_numerics.

Reply:
OK, I see the difference of "skip just those extremal tests instead of skipping all of the historic test_reference_numerics." But I still don't think we should enable this nan is nan compare only for the failing tests. Because, although `sin` is not failing like `asin`, I still don't think `sin(x) = 2 + nan j` vs `sin(x) = 3 - nan j` one is more correct than the other. We can have a new field in `relaxed_complex_nan` in `OpInfo`, but most values of that field should be `True` instead of `False`.

Reply:
I hear you, and you're not wrong, but let's start with this more focused fix and only add this setting where necessary for now.

Reply:
There are actually a lot of tests failing due to the same reason
```
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_extremal_acosh_cuda_complex128 - AssertionError: False is not true : Scalars failed to compare as equal! Comparing the imaginary ...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_extremal_acosh_cuda_complex64 - AssertionError: False is not true : Scalars failed to compare as equal! Comparing the imaginary p...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_extremal_asin_cuda_complex128 - AssertionError: False is not true : Scalars failed to compare as equal! Comparing the imaginary p...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_extremal_asin_cuda_complex64 - AssertionError: False is not true : Scalars failed to compare as equal! Comparing the imaginary pa...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_extremal_asinh_cuda_complex128 - AssertionError: False is not true : Scalars failed to compare as equal! Comparing the real part ...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_extremal_asinh_cuda_complex64 - AssertionError: False is not true : Scalars failed to compare as equal! Comparing the real part -...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_extremal_log10_cuda_complex128 - AssertionError: False is not true : Scalars failed to compare as equal! Comparing the real part ...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_extremal_log10_cuda_complex64 - AssertionError: False is not true : Scalars failed to compare as equal! Comparing the real part n...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_extremal_log_cuda_complex128 - AssertionError: False is not true : Scalars failed to compare as equal! Comparing the real part na...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_extremal_log_cuda_complex64 - AssertionError: False is not true : Scalars failed to compare as equal! Comparing the real part nan...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_extremal_tan_cuda_complex128 - AssertionError: False is not true : Scalars failed to compare as equal! Comparing the imaginary pa...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_extremal_tan_cuda_complex64 - AssertionError: False is not true : Scalars failed to compare as equal! Comparing the imaginary par...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_extremal_tanh_cuda_complex128 - AssertionError: False is not true : Scalars failed to compare as equal! Comparing the real part n...
FAILED test/test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_extremal_tanh_cuda_complex64 - AssertionError: False is not true : Scalars failed to compare as equal! Comparing the real part na...
```

Reply:
Cool!